### DISCUSSION TITLE
Add `--init` or `--config` Command for Shuru Configuration Initialization

### DISCUSSION THREAD

**Discussion by harshdoesdev** [shuru-project/shuru#26] [Ideas]
# Add `--init` or `--config` Command for Shuru Configuration Initialization

We’re looking to add a `--init` or `--config` command to the Shuru task runner and version manager to improve the initial setup process for users. This command will help automate the creation of a `shuru.toml` file for new projects.

### 1. Check for Existing Configuration:
- The command should check for an existing shuru.toml file in the current directory.
- If the file exists, the user should be warned and asked whether they want to overwrite the current configuration. If they choose not to, the command should exit.

### 2. Project Type Detection:
- The command should detect the project type by looking for a `package.json` or other relevant project files (supporting Python and Node.js).
- If a `package.json` is found, it should read the scripts section and migrate those commands into `shuru.toml` as tasks. This can be handled using `serde_json` for processing the JSON data.

### 3. Configuration (`shuru.toml`) File Generation
- If no relevant project files (like `package.json`) are found, and the project seems empty, the command should prompt the user to select options for generating a basic configuration.
	- Ask the user if they want to initialize the project for Node.js or Python.
	- Prompt for selecting the version of Node.js or Python they wish to use.
	- Optionally, offer to initialize the project with basic commands for tasks like running the app, testing, or building.
	- Offer to initialize Git if no `.git` directory is found, asking if the user wants to create a `.gitignore` file.
	- The resulting `shuru.toml` should contain default tasks related to the selected options (e.g., running a basic Node.js server or Python script), and include version management information where applicable.

## Context

- `src/commands.rs`: This file contains the functions for handling various CLI commands, such as updating versions, listing tasks, clearing cache, and generating shell completions. The new function for the `--init` (or `--config`) command should also be defined here to handle initializing the `shuru.toml` configuration file.
- `src/config.rs`: This file defines how tasks and version configurations are managed, with structures for `Config` and `TaskConfig`, including validation methods. This will be relevant for generating the default `shuru.toml` file.

## Benefits
- This feature will make it easier for new users to get started with Shuru by automatically generating a project-specific config file.
- It reduces the manual effort required to create a `shuru.toml`, making the tool more user-friendly.

## Example Usage

```sh
shuru --init [PROJECT_DIR]

# or shuru --config [PROJECT_DIR]
```

## Acceptance Criteria

- Implement a check for the existence of `shuru.toml`.
- Prompt the user for overwriting if `shuru.toml` exists.
- Detect package.json or similar files to identify the project type.
- Read the scripts section from `package.json` and convert these to tasks in shuru.toml using `serde_json`.
- If no relevant project files are found, prompt the user to initialize the project and generate a basic `shuru.toml` configuration, including project type (Node.js or Python), version selection, basic tasks, and Git initialization.
- Write unit tests to ensure functionality works as expected.

Let me know if you have any questions!

**Comment by Aditya-PS-05** [comment]
Please assign it to me under hacktoberfest label.

**Comment by harshdoesdev** [comment]
Hi @Aditya-PS-05, this issue is not part of Hacktoberfest. Thank you for showing interest! Feel free to participate in the discussion and open a PR. Once the specification is ready, we can proceed with the implementation and review any contributions.

**Comment by prxjwal** [comment]
I would like to work on this idea

**Reply by harshdoesdev** [reply]
Great! Please take a look at the spec, and let’s discuss and finalize the details.

CC: @airvzxf, @Aditya-PS-05

**Reply by harshdoesdev** [reply]
First problem we have to solve is that it’s easy to identify a Node.js project by checking for `package.json`. How can we do the same for Python? Checking for these?:

- requirements.txt
- .python-version
- Pipfile
- setup.py
- venv/ directory

**Reply by airvzxf** [reply]
> First problem we have to solve is that it’s easy to identify a Node.js project by checking for `package.json`. How can we do the same for Python? Checking for these?:
> 
>     * requirements.txt
> 
>     * .python-version
> 
>     * Pipfile
> 
>     * setup.py
> 
>     * venv/ directory

Could you provide a description of the representative command for each one of these files/folders:
* .python-version ➟ Insert the version in the _shuru.toml_
* requirements.txt ➟ `pip install`
* Pipfile ➟ `pip install`
* setup.py ➟ ???
* venv/ directory ➟ ???

**Reply by harshdoesdev** [reply]
Thanks, this sums it up. For the last two, these were for adding relevant commands in the `shuru.toml` files, but it can be left to the user to add them in the initial implementation.

**Comment by airvzxf** [comment]
# Sub-feature proposal

## General proposal

I would like to add a sub-feature to this discussion. Configure the technical installation of Python, possibly applicable for Node.

## Proposal for Python

This means that I can indicate which arguments I am going to add when I run the './configure' command. For example, I want to use all 8 cores of my computer when compiling instead of just one because the Python compilation is very slow. With this, I could also add arguments such as Python optimizations.

In this dynamic, I suggest that these settings are always set by default and are overwritten when the Python installation is installing.

## Proposal for the installation prompt

In the installation prompt, we can add advanced or simple installation. The simple one would basically do what you propose. The advanced one would allow you to select all these values (build cores, optimizations), obviously leaving some by default.

## Configuration file proposal

With this trend towards scalability and a bit of SaaS (software as a service), the best thing would be to be able to create a file where all these configurations are taken and created, without the need to enter manual entries in the prompt.

**Reply by harshdoesdev** [reply]
Awesome! We are going to put these configs in the shuru.toml file. Just like how we can specify the platform along with the version for Node.js:

```toml
node = { version = "v16.14.0", platform = "darwin-arm64" } 
```

Similarly, we can specify the configuration for Python.

**Comment by airvzxf** [comment]
# Proposed sub-feature

## General proposal

Add the option to install autocomplete. Taking advantage of the fact that we are already doing everything via prompt, we could add an option that says if you want to automatically add the autocomplete environment.

Checking around, I saw that you can do something like this: “https://github.com/shuru-project/shuru/discussions/11”.
```bash
shuru --completions zsh > /Users/<username>/.oh-my-zsh/custom/completions/_shuru
```

It has already been discussed that it is better to let the user decide this part, but I insist that it should be good for them to have the autocomplete configured from the beginning so that they can use it. To this day, I still cannot use the autocomplete because I am too lazy to install it.

**Reply by harshdoesdev** [reply]
The completions script is not a project-level setting but a system-level setting. Maybe we can provide this option in the install.sh script or add a new command for it?

**Comment by airvzxf** [comment]
# Main feature

## General

Overall, I like the idea. Honestly, at first glance, I don't see anything else that can be added, modified, or removed. I think all those comments will come when we are working on this Epic/Future.

## Flowchart (architectural design)

I suggest we make a flow diaphragm or something like that so we don't get lost along the way. I've been reading, and I think that when we are programming, we are going to get confused. Possibly, it can be easily put together from whatever it seems, and we can do it quickly and cleanly. This is just a personal comment, although the truth is that I'm too lazy to make flowcharts or states.

## Work method

If this feature is intended for multiple users to program at the same time, I think that a way of working should be well-defined so as not to overlap the changes in the files. Which in Git is always a headache.

## Detect Python projects

> First problem we have to solve is that it’s easy to identify a Node.js project by checking for package.json. How can we do the same for Python? Checking for these?:
>    requirements.txt
>    .python-version
>    Pipfile
>    setup.py
>    venv/ directory



I really find it difficult to achieve the satisfaction that it automatically searches for which project it is and makes a configuration. Let's stick to 2 or 3 very direct ones that don't cause confusion and don't clash. I would start with these two:

- requirements.txt
- .python-version

**Reply by harshdoesdev** [reply]
Agreed. Unlike in Node.js, where we can directly copy the commands by reading the `.scripts` section of `package.json`, Python doesn’t have a standardized way to do this. So, in the default `shuru.toml` generated for Python, we will only include the version specified in `.python-version`. Additionally, if there is a `requirements.txt` file, we can add the `pip install` command as a task.

**Reply by harshdoesdev** [reply]
Yes, we can definitely create a flowchart to track the flow, as it will involve multiple configurations and be a separate module.

**Comment by Aditya-PS-05** [comment]
Can we try something like this?
```
shuru --init
# Starts project initialization

shuru --config list
# Lists current configuration settings or may be the dependencies used in the project.
```

**Reply by harshdoesdev** [reply]
Awesome! Yeah, using the `--config` command for this purpose is better. It’s similar to AWS and Git CLI.

After giving it some thought, I believe `shuru --project [dir]` would be a better command name for starting a project since the word “Shuru” literally means “start” in many languages. What do you guys think?

**Reply by Aditya-PS-05** [reply]
what will be the use of `--init` then?

Will this work?
```
shuru --project [dir]
   --python@version  # creates a blank python project
   --node@version # creates a blank node.js project

shuru --init
# Starts project initialization

shuru --config list
# Lists current configuration settings or may be the dependencies used in the project.
```

**Reply by harshdoesdev** [reply]
Actually, I was thinking of renaming the `shuru --init` command to `shuru --project`, but the method you suggested for passing specific versions of Python and Node looks awesome.

**Reply by airvzxf** [reply]
## PROPOSAL

I created a Matrix to check all the possibilities, then we can clean the arguments and sub-arguments. Currently, I think a matrix is better than a flow, with this matrix we cloud have better visibility.

## MATRIX
The spreadsheet is open to edit by anyone, later I'll close to view only, to prevent hacks. You are welcome to add, edit or remove columns or rows.
https://docs.google.com/spreadsheets/d/1ljI_ngmD-YNEVHxNaGJat3y6HY3y4cHwRajxWKp1uUQ/edit

![Matrix in general](https://github.com/user-attachments/assets/bf9046fd-0986-4820-814f-effad157f9ee)

## HOW TO FILL
The dynamic that I propose: I created tabs with your GitHub IDs to everyone fill as you think is the best way to have these arguments.

![Tabs of the matrix](https://github.com/user-attachments/assets/440195a5-0aeb-4dc7-9a53-02f9456cc9e2)


## COLUMNS

* **Target path:** The directory path of the project that we want to “open” or set up.
* **Exists folder:** Yes, if the folder exists in the Operative System.
* **Empty folder:** Yes, if the folder exists, but it is “empty”, without Shuru, Python or Node files.
* **Python:** It is, or it should have a Python project.
* **Node:** It is, or it should have a Node project.
* **Error:** You can fill with Yes, No, or a description of the error that should be thrown this command for this case.
* **Command:** For all is `./shuru`.
* **Main argument:** The main argument used for this set up.
* **sub-arguments:** The subsequent arguments to refine this set up.

**Reply by harshdoesdev** [reply]
Thanks a ton!

**Reply by harshdoesdev** [reply]
Hi @airvzxf, I don’t think I have edit access to the document, but I reviewed your tab, and my decision aligns with the configuration you proposed.

**Reply by airvzxf** [reply]
Sorry, it was per default view only. I added as an editor for anyone.
https://docs.google.com/spreadsheets/d/1ljI_ngmD-YNEVHxNaGJat3y6HY3y4cHwRajxWKp1uUQ/edit?usp=sharing

**Comment by Aditya-PS-05** [comment]
Updated the Spreadsheet!. I still have some confusions regarding this.

**Reply by airvzxf** [reply]
What is your confusions?

**Reply by Aditya-PS-05** [reply]
> What is your confusions?

![Screenshot from 2024-10-13 20-33-49](https://github.com/user-attachments/assets/69e43291-439d-4a47-a340-70049ea1575f)
In this specific case, if the folder is not empty. Then two cases: 
1.) If one of python and node project is defined. 
2.) Suppose if neither python nor node is not defined. We need to throw a error right.

**Reply by harshdoesdev** [reply]
If no version is specified by the user, we can look for files created by other version managers, such as `.python-version` or `.nvmrc`, and read the versions from there. We could then prompt the user to see if they would like to copy these into the `shuru.toml` file.

**Reply by airvzxf** [reply]
This matrix exercise paid off, I'm happy 😄.

The @Aditya-PS-05 comment opened a new meta-universe (joke). Because we need to define first what is the follow steps after executing the commands with the arguments, and I am avoiding the obviously steps. **In summary,** when the **user executes** those commands: should it **automatically create or set up the project**? Or, should it **show the installation prompt** with the default options regarding the arguments?

Only for clarification, I assumed that the prompt installation keeps in our plans.

> ### 3. Configuration (`shuru.toml`) File Generation
> * If no relevant project files (like `package.json`) are found, and the project seems empty, **the command should prompt the user to select options for generating a basic configuration.**
>   * Ask the user if they want to initialize the project for Node.js or Python.
>   * **Prompt for selecting the version of Node.js or Python they wish to use.**
>   * Optionally, offer to initialize the project with basic commands for tasks like running the app, testing, or building.
>   * Offer to initialize Git if no `.git` directory is found, asking if the user wants to create a `.gitignore` file.
>   * The resulting `shuru.toml` should contain default tasks related to the selected options (e.g., running a basic Node.js server or Python script), and include version management information where applicable.

---
Yes, for this specific case, there are two subcases (three in reality, if we split Python and node, but as you mentioned, is good).

> In this specific case, if the folder is not empty. Then two cases:
> 1.) If one of python and node project is defined.
> 2.) Suppose if neither python nor node is not defined. We need to throw an error right.

Answering directly your question, Aditya. In the case 2), As Harash mentioned, it could show the prompt and asks for what type of project is looking for, Python or Node. In another option, as you mentioned, it could throw an error about specifying the type of the project: Python or Node.

We need to define how it will work for this command and arguments, and the future commands. Prompts or direct executions.

**Reply by harshdoesdev** [reply]
Exactly! Adding to your last point, if there are existing versions defined in `.nvmrc` files or `.python-version`, we can use those as default values while promoting the user or show a warning if the user is specifying a different version via arguments to the cli. Like, if you have specified node v16 in your `.nvmrc` file, but while passing the arguments you have passed v18. It is not going to be a prompt, just a warning.

**Reply by airvzxf** [reply]
And what happens in these cases, which are technically the same:

Target path | Exists folder | Empty folder | Python | Node
-- | -- | -- | -- | --
Yes | Yes | Yes | N/A | N/A
Yes | Yes | No | N/A | N/A

**Reply by harshdoesdev** [reply]
Also, I am opposed to throwing any errors in this case. If it is a new/empty project and the user did not pass or did not select any versions, shuru shall just create a `shuru.toml` file with comments to aid the user to set it up manually. Like, create an empty `[versions]` table and below it add comments showing examples of adding Node and Python versions. What is your opinion on this?

**Reply by harshdoesdev** [reply]
I think my above comment answers this.

**Reply by Aditya-PS-05** [reply]
> Also, I am opposed to throwing any errors in this case. If it is a new/empty project and the user did not pass or did not select any versions, shuru shall just create a `shuru.toml` file with comments to aid the user to set it up manually. Like, create an empty `[versions]` table and below it add comments showing examples of adding Node and Python versions. What is your opinion on this?

Now I also think, creating a file is much better than throwing an error.

**Comment by airvzxf** [comment]
## Other discussion: Missing path.

This discussion is about for the case where the user didn't specify the path.

We need to clarify if the argument for `--project` requires the path or is it optional.
* `--project` | `--project [PATH]` | If a path is not specified, take the current path or some fixed path (need to check with team).
* `--project ~/my-project` | `--project PATH`

## Cases

I added some cases below to provide visibility.

### Case 1.1)

#### Set up

I am on my terminal, type: `cd ~/my-project`, and execute `shuru --project`.

#### Expectation

I imagine that it will check in `~/my-project`. If the Python or Node files existed, it created the `~/my-projectshuru.toml` otherwise it creates a template for the same file. Discussed in above post.

### Case 1.2)

#### Set up

I am on my terminal, specially in my home folder `~/`, then execute `shuru --project`.

#### Expectation

I imagine that it will check in `~/`. If the Python or Node files existed, it created the `~/shuru.toml` otherwise it creates a template for the same file. Discussed in above post.

### Case 2)

#### Set up

I am on my terminal, specially in my home folder `~/`, then execute `shuru --project ~/my-project`.

#### Expectation

I imagine that it will check in `~/my-project`. If the Python or Node files existed, it created the `~/my-project/shuru.toml` otherwise it creates a template for the same file. Discussed in above post.
